### PR TITLE
Remove duplicate fork definitions

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

`type fork uint8` seems to be very popular... Moving this into a common package and sharing where possible.

## How this works

Adds an `upgradetest` package and exposes common functionality that existing VMs re-implement multiple times.

## How this was tested

- [X] CI

### CODE REVIEW

File: vms/avm/environment_test.go [hunk 1]
```diff
@@ -42,14 +40,7 @@ import (
 	keystoreutils "github.com/ava-labs/avalanchego/vms/components/keystore"
 )
 
-type fork uint8
```
Comment (StephenButtolph) [inline vms/avm/environment_test.go:45]: 1

File: vms/platformvm/block/builder/helpers_test.go [hunk 2]
```diff
@@ -97,8 +88,6 @@ func init() {
 	}
 }
 
-type fork uint8
```
Comment (StephenButtolph) [inline vms/platformvm/block/builder/helpers_test.go:100]: 2

File: vms/platformvm/block/executor/helpers_test.go [hunk 3]
```diff
@@ -100,8 +93,6 @@ func init() {
 
 type stakerStatus uint
 
-type fork uint8
```
Comment (StephenButtolph) [inline vms/platformvm/block/executor/helpers_test.go:103]: 3

File: vms/platformvm/txs/executor/helpers_test.go [hunk 4]
```diff
@@ -90,8 +83,6 @@ func init() {
 	}
 }
 
-type fork uint8
```
Comment (StephenButtolph) [inline vms/platformvm/txs/executor/helpers_test.go:93]: 4

File: vms/platformvm/vm_test.go [hunk 5]
```diff
@@ -162,8 +150,6 @@ func init() {
 	}
 }
 
-type fork uint8
```
Comment (StephenButtolph) [inline vms/platformvm/vm_test.go:165]: 5

File: upgrade/upgradetest/config.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	// Initialize all forks to be unscheduled
+	SetConfigTimesTo(&c, Latest, upgrade.UnscheduledActivationTime)
+	// Schedule the requested forks at the provided upgrade time
+	SetConfigTimesTo(&c, fork, upgradeTime)
+	return c
+}
+
+// SetConfigTimesTo sets the upgrade time of the provided fork, and all prior
+// forks, to the provided upgradeTime.
+func SetConfigTimesTo(c *upgrade.Config, fork Fork, upgradeTime time.Time) {
```
Comment (StephenButtolph) [inline upgrade/upgradetest/config.go:35]: The signature looks like this to align with a pointer receiver method on the `Config` type (which can't be done here because this is a testing package)

File: upgrade/upgradetest/config.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/ava-labs/avalanchego/upgrade"
+)
+
+// GetConfig returns an upgrade config with the provided fork scheduled to have
+// been initially activated and all other forks to be unscheduled.
+func GetConfig(fork Fork) upgrade.Config {
+	return GetConfigWithUpgradeTime(fork, upgrade.InitiallyActiveTime)
+}
+
+// GetConfig returns an upgrade config with the provided fork scheduled to be
```
Comment (dhrubabasu) [inline upgrade/upgradetest/config.go:19]: preempting the sybils

```suggestion
// GetConfigWithUpgradeTime returns an upgrade config with the provided fork scheduled to be
```

File: upgrade/upgradetest/config.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	SetTimesTo(&c, Latest, upgrade.UnscheduledActivationTime)
+	// Schedule the requested forks at the provided upgrade time
+	SetTimesTo(&c, fork, upgradeTime)
+	return c
+}
+
+// SetTimesTo sets the upgrade time of the provided fork, and all prior forks,
+// to the provided upgradeTime.
+func SetTimesTo(c *upgrade.Config, fork Fork, upgradeTime time.Time) {
+	switch fork {
```
Comment (ARR4N) [inline upgrade/upgradetest/config.go:37]: Elegant 😍

File: vms/platformvm/block/builder/helpers_test.go [hunk 9]
```diff
@@ -303,8 +292,16 @@ func defaultState(
...
 }
 
-func defaultConfig(t *testing.T, f fork) *config.Config {
-	c := &config.Config{
+func defaultConfig(f upgradetest.Fork) *config.Config {
+	upgrades := upgradetest.GetConfigWithUpgradeTime(f, time.Time{})
+	// This package neglects fork ordering
+	upgradetest.SetTimesTo(
+		&upgrades,
+		min(f, upgradetest.ApricotPhase5),
```
Comment (ARR4N) [inline vms/platformvm/block/builder/helpers_test.go:300]: Why Apricot 5?

File: vms/platformvm/block/executor/verifier_test.go [hunk 10]
```diff
@@ -61,9 +61,7 @@ func TestVerifierVisitProposalBlock(t *testing.T) {
 	manager := &manager{
 		txExecutorBackend: &executor.Backend{
 			Config: &config.Config{
-				UpgradeConfig: upgrade.Config{
-					BanffTime: mockable.MaxTime, // banff is not activated
-				},
+				UpgradeConfig: upgradetest.GetConfig(upgradetest.ApricotPhasePost6),
```
Comment (ARR4N) [inline vms/platformvm/block/executor/verifier_test.go:64]: I really like this pattern. It's much more declarative around intent / effect.

File: vms/proposervm/vm_test.go [hunk 11]
```diff
@@ -808,11 +809,7 @@ func TestExpiredBuildBlock(t *testing.T) {
 	proVM := New(
 		coreVM,
 		Config{
-			Upgrades: upgrade.Config{
-				ApricotPhase4Time:            time.Time{},
-				ApricotPhase4MinPChainHeight: 0,
-				DurangoTime:                  mockable.MaxTime,
-			},
+			Upgrades:            upgradetest.GetConfigWithUpgradeTime(upgradetest.ApricotPhase4, time.Time{}),
```
Comment (ARR4N) [inline vms/proposervm/vm_test.go:812]: Does the (Go) zero-value `time.Time{}` have a special meaning?

And the Unix zero-value? Is this effectively the same as using `upgrade.InitiallyActiveTime`?

**Review by ARR4N — APPROVED** [review]
I'm concerned that the philosophers will now die of starvation.

All comments are for my own learning.

File: vms/platformvm/block/builder/helpers_test.go (hunk 9 above)
Comment (StephenButtolph) [reply vms/platformvm/block/builder/helpers_test.go:300]: This was done just to align with the previous code... This really should not do this (begs into question the tests that use this tbh). But without this hack - the tests fail. Hoping to keep this PR small.

File: vms/proposervm/vm_test.go (hunk 11 above)
Comment (StephenButtolph) [reply vms/proposervm/vm_test.go:812]: These values seem to be depended on in the tests that use this code... I'm hoping that we can remove the reliance on `time.Time{}` and `time.Unix(0, 0)` as we continue to fix these tests... But it felt out of scope for this PR...

TL;DR: I don't know... Changing them breaks the tests... The tests should be improved in a later PR.

File: upgrade/upgradetest/config.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+func GetConfig(fork Fork) upgrade.Config {
+	return GetConfigWithUpgradeTime(fork, upgrade.InitiallyActiveTime)
+}
+
+// GetConfigWithUpgradeTime returns an upgrade config with the provided fork
+// scheduled to be activated at the provided upgradeTime and all other forks to
+// be unscheduled.
+func GetConfigWithUpgradeTime(fork Fork, upgradeTime time.Time) upgrade.Config {
+	c := upgrade.Config{
+		ApricotPhase4MinPChainHeight: 0,
```
Comment (yacovm) [inline upgrade/upgradetest/config.go:24]: this is the zero value so why explicitly set it?

File: vms/proposervm/vm_test.go [hunk 13]
```diff
@@ -1600,11 +1589,7 @@ func TestRejectedHeightNotIndexed(t *testing.T) {
 	proVM := New(
 		coreVM,
 		Config{
-			Upgrades: upgrade.Config{
-				ApricotPhase4Time:            time.Unix(0, 0),
-				ApricotPhase4MinPChainHeight: 0,
-				DurangoTime:                  time.Unix(0, 0),
-			},
+			Upgrades:            upgradetest.GetConfigWithUpgradeTime(upgradetest.Latest, time.Unix(0, 0)),
```
Comment (yacovm) [inline vms/proposervm/vm_test.go:1592]: I know it's no the same, as one is 1970 and the other is just 0, but, why the discrepancy between time.Unix(0,0) and time.Time? Won't this test work with time.Time{} ?

**Review by yacovm — APPROVED** [review]

File: upgrade/upgradetest/config.go (hunk 12 above)
Comment (StephenButtolph) [reply upgrade/upgradetest/config.go:24]: We could remove this - I felt like it was a useful shout-out... But I'm fine deleting it.

Comment (StephenButtolph) [reply upgrade/upgradetest/config.go]: removed.

**Review by dhrubabasu — APPROVED** [review]
